app/NounsServices/GraphQL/GraphQL.swift

```diff
@@ -22,36 +23,97 @@ enum GraphError: Error {
+      true).first!
+    let documentsURL = URL(fileURLWithPath: documentsPath)
+    let sqliteFileURL = documentsURL.appendingPathComponent("nouns_test.sqlite")
+    
+    guard let cache = try? SQLiteNormalizedCache(fileURL: sqliteFileURL) else {
+      // If SQLite fails for whatever reason (file corrupt, deleted), default to memory cache
+      return ApolloStore(cache: InMemoryNormalizedCache())
+    }
+    
+    let store = ApolloStore(cache: cache)
+    return store
+  }()
+  
+  /// A web socket transport to use for subscriptions
+  private lazy var webSocketTransport: WebSocketTransport = {
+    let request = URLRequest(url: websocketURL)
+    let webSocketClient = WebSocket(request: request)
+    return WebSocketTransport(websocket: webSocketClient, store: store)
+  }()
+  
+  /// An HTTP transport to use for queries and mutations
+  private lazy var httpTransport: RequestChainNetworkTransport = {
+    return RequestChainNetworkTransport(interceptorProvider: DefaultInterceptorProvider(store: store), endpointURL: url)
+  }()
+  
+  /// A split network transport to allow the use of both of the above
+  /// transports through a single `NetworkTransport` instance.
+  private lazy var splitNetworkTransport = SplitNetworkTransport(
+    uploadingNetworkTransport: self.httpTransport,
+    webSocketNetworkTransport: self.webSocketTransport
+  )
+  
+  private(set) lazy var apollo: ApolloClient = ApolloClient(networkTransport: splitNetworkTransport, store: store)
+  
+  public init(httpURL: URL, websocketURL: URL) {
+    self.url = httpURL
+    self.websocketURL = websocketURL
+    
+    self.apollo.cacheKeyForObject = { $0["id"] }
   }
   
   func fetch<T>(_ query: T) -> AnyPublisher<T.Data, Error> where T: GraphQLQuery {
-    let future = Future<T.Data, Error> { promise in
-      self.apollo.fetch(query: query) { result in
-        if let errors = try? result.get().errors {
-          promise(.failure(GraphError.errors(errors)))
+    let subject = PassthroughSubject<T.Data, Error>()
+    
+    self.apollo.fetch(query: query, cachePolicy: .returnCacheDataAndFetch) { result in
+      if let errors = try? result.get().errors {
+        subject.send(completion: .failure(GraphError.errors(errors)))
+        return
+      }
+      
+      do {
+        guard let data = try result.get().data else {
+          subject.send(completion: .failure(GraphError.emptyResponse))
           return
         }
-        
-        do {
-          guard let data = try result.get().data else {
-            promise(.failure(GraphError.emptyResponse))
-            return
-          }
-          promise(.success(data))
-        } catch let error {
-          promise(.failure(GraphError.retrieveError(error)))
-        }
+        subject.send(data)
+      } catch let error {
+        subject.send(completion: .failure(GraphError.retrieveError(error)))
       }
     }
     
-    return future.eraseToAnyPublisher()
+    return subject.eraseToAnyPublisher()
```

Comment:
The request does return a `source` either being `cache` or `server`. We can return a completion if it's from the server as that'll be the most up to date information anyways. If it fails (no internet for example), then that error is already handled and it will send a failure completion